File: data_structures/6_go/f_graphs.go
package main

import (
	"fmt"
	"math"
)

type Edge struct {
	destination *GraphNode
	weight      *float64
}

type GraphNode struct {
	value int
	edges []*Edge
}

type Graph struct {
	nodes []*GraphNode
}

// NewGraphNode creates a new graph node with the given value.
func NewGraphNode(value int) *GraphNode {
	return &GraphNode{
		value: value,
		edges: []*Edge{},
	}
}

// NewGraph creates a new graph.
func NewGraph() *Graph {
	return &Graph{
		nodes: []*GraphNode{},
	}
}

// AddNode adds a node to the graph.
func (g *Graph) AddNode(value int) {
	node := NewGraphNode(value)
	g.nodes = append(g.nodes, node)
}

// RemoveNode removes a node from the graph.
func (g *Graph) RemoveNode(value int) {
	var newNodes []*GraphNode

	for _, node := range g.nodes {
		if node.value != value {
			newNodes = append(newNodes, node)
		}
	}

	g.nodes = newNodes

	for _, node := range g.nodes {
		var newEdges []*Edge

		for _, edge := range node.edges {
			if edge.destination.value != value {
				newEdges = append(newEdges, edge)
			}
		}

		node.edges = newEdges
	}
}

// AddEdge adds an edge between two nodes in the graph.
func (g *Graph) AddEdge(sourceValue, destinationValue int, weight *float64) {
	sourceNode := g.getNode(sourceValue)
	destinationNode := g.getNode(destinationValue)

	if sourceNode != nil && destinationNode != nil {
		edge := &Edge{
			destination: destinationNode,
			weight:      weight,
		}
		sourceNode.edges = append(sourceNode.edges, edge)
	}
}

// RemoveEdge removes an edge between two nodes in the graph.
func (g *Graph) RemoveEdge(sourceValue, destinationValue int) {
	sourceNode := g.getNode(sourceValue)

	if sourceNode != nil {
		var newEdges []*Edge

		for _, edge := range sourceNode.edges {
			if edge.destination.value != destinationValue {
				newEdges = append(newEdges, edge)
			}
		}

		sourceNode.edges = newEdges
	}
}

// DepthFirstSearch performs a depth-first search starting from a specific node value.
func (g *Graph) DepthFirstSearch(startValue int) []int {
	startNode := g.getNode(startValue)
	visited := make(map[int]bool)
	var result []int

	if startNode != nil {
		g.depthFirstSearchRec(startNode, visited, &result)
	}

	return result
}

func (g *Graph) depthFirstSearchRec(node *GraphNode, visited map[int]bool, result *[]int) {
	if visited[node.value] {
		return
	}

	visited[node.value] = true
	*result = append(*result, node.value)

	for _, edge := range node.edges {
		g.depthFirstSearchRec(edge.destination, visited, result)
	}
}

// BreadthFirstSearch performs a breadth-first search starting from a specific node value.
func (g *Graph) BreadthFirstSearch(startValue int) []int {
	startNode := g.getNode(startValue)
	visited := make(map[int]bool)
	var result []int
	var queue []*GraphNode

	if startNode != nil {
		queue = append(queue, startNode)
	}

	for len(queue) > 0 {
		currentNode := queue[0]
		queue = queue[1:]

		if visited[currentNode.value] {
			continue
		}

		visited[currentNode.value] = true
		result = append(result, currentNode.value)

		for _, edge := range currentNode.edges {
			queue = append(queue, edge.destination)
		}
	}

	return result
}

// Utility function to get a node by its value.
func (g *Graph) getNode(value int) *GraphNode {
	for _, node := range g.nodes {
		if node.value == value {
			return node
		}
	}
	return nil
}

// IsCyclic checks if the graph is cyclic.
func (g *Graph) IsCyclic() bool {
	visited := make(map[int]bool)
	stack := make(map[int]bool)

	for _, node := range g.nodes {
		if g.isCyclicHelper(node, visited, stack) {
			return true
		}
	}

	return false
}

func (g *Graph) isCyclicHelper(node *GraphNode, visited, stack map[int]bool) bool {
	if stack[node.value] {
		return true
	}

	if visited[node.value] {
		return false
	}

	visited[node.value] = true
	stack[node.value] = true

	for _, edge := range node.edges {
		if g.isCyclicHelper(edge.destination, visited, stack) {
			return true
		}
	}

	stack[node.value] = false

	return false
}

func edgeWeightOrZero(weight *float64) float64 {
	if weight != nil {
		return *weight
	}
	return 0
}

// ShortestPath finds the shortest path between two nodes using Dijkstra's algorithm.
func (g *Graph) ShortestPath(startValue, endValue int) []int {
	startNode := g.getNode(startValue)
	endNode := g.getNode(endValue)

	if startNode == nil || endNode == nil {
		return nil
	}

	distances := make(map[int]float64)
	previousNodes := make(map[int]int)
	unvisited := make(map[int]bool)

	for _, node := range g.nodes {
		distances[node.value] = math.Inf(1)
		previousNodes[node.value] = -1
		unvisited[node.value] = true
	}

	distances[startNode.value] = 0

	for len(unvisited) > 0 {
		minNodeValue := g.getMinDistanceNode(distances, unvisited)
		if minNodeValue == -1 {
			break
		}

		minNode := g.getNode(minNodeValue)
		delete(unvisited, minNodeValue)

		for _, edge := range minNode.edges {
			neighborValue := edge.destination.value
			alt := distances[minNodeValue] + edgeWeightOrZero(edge.weight)

			if alt < distances[neighborValue] {
				distances[neighborValue] = alt
				previousNodes[neighborValue] = minNodeValue
			}
		}
	}

	// Build the path from endNode to startNode
	path := []int{endValue}
	current := endValue
	for previousNodes[current] != -1 {
		current = previousNodes[current]
		path = append([]int{current}, path...)
	}

	return path
}

func (g *Graph) getMinDistanceNode(distances map[int]float64, unvisited map[int]bool) int {
	minDist := math.Inf(1)
	minNode := -1

	for nodeValue, dist := range distances {
		if unvisited[nodeValue] && dist < minDist {
			minDist = dist
			minNode = nodeValue
		}
	}

	return minNode
}

// printGraph prints the graph in adjacency list representation.
func (g *Graph) printGraph() {
	for _, node := range g.nodes {
		fmt.Printf("Node %d:", node.value)
		for _, edge := range node.edges {
			fmt.Printf(" -> %d", edge.destination.value)
		}
		fmt.Println()
	}
}

File: data_structures/3_python/f_graphs.py
from typing import TypeVar, List, Optional, Dict, Set

T = TypeVar('T')


class Edge:
    def __init__(self, destination: 'GraphNode', weight: Optional[float] = None):
        self.destination = destination
        self.weight = weight


class GraphNode:
    def __init__(self, value: T):
        self.value = value
        self.edges: List[Edge] = []


class Graph:
    def __init__(self):
        self.nodes: List[GraphNode] = []

    # MARK: - Node Operations

    def add_node(self, value: T):
        new_node = GraphNode(value)
        self.nodes.append(new_node)

    def remove_node(self, value: T):
        self.nodes = [node for node in self.nodes if node.value != value]
        for node in self.nodes:
            node.edges = [edge for edge in node.edges if edge.destination.value != value]

    # MARK: - Edge Operations

    def add_edge(self, source_value: T, destination_value: T, weight: Optional[float] = None):
        source_node = self.get_node(source_value)
        destination_node = self.get_node(destination_value)

        if source_node and destination_node:
            edge = Edge(destination_node, weight)
            source_node.edges.append(edge)

    def remove_edge(self, source_value: T, destination_value: T):
        source_node = self.get_node(source_value)

        if source_node:
            source_node.edges = [edge for edge in source_node.edges if edge.destination.value != destination_value]

    # MARK: - Search

    def depth_first_search(self, start_value: T) -> List[T]:
        start_node = self.get_node(start_value)
        visited = set()
        result = []

        if start_node:
            self._depth_first_search_rec(start_node, visited, result)

        return result

    def _depth_first_search_rec(self, node: GraphNode, visited: Set[T], result: List[T]):
        if node.value in visited:
            return

        visited.add(node.value)
        result.append(node.value)

        for edge in node.edges:
            self._depth_first_search_rec(edge.destination, visited, result)

    def breadth_first_search(self, start_value: T) -> List[T]:
        start_node = self.get_node(start_value)
        visited = set()
        result = []
        queue = [start_node] if start_node else []

        while queue:
            current_node = queue.pop(0)

            if current_node.value in visited:
                continue

            visited.add(current_node.value)
            result.append(current_node.value)

            for edge in current_node.edges:
                queue.append(edge.destination)

        return result

    # MARK: - Utility

    def get_node(self, value: T) -> Optional[GraphNode]:
        for node in self.nodes:
            if node.value == value:
                return node
        return None

    # MARK: - Check Graph Properties

    def is_cyclic(self) -> bool:
        visited = set()
        stack = set()

        for node in self.nodes:
            if self._is_cyclic_helper(node, visited, stack):
                return True

        return False

    def _is_cyclic_helper(self, node: GraphNode, visited: Set[T], stack: Set[T]) -> bool:
        if node.value in stack:
            return True

        if node.value in visited:
            return False

        visited.add(node.value)
        stack.add(node.value)

        for edge in node.edges:
            if self._is_cyclic_helper(edge.destination, visited, stack):
                return True

        stack.remove(node.value)
        return False

    # MARK: - Shortest Path

    def shortest_path(self, start_value: T, end_value: T) -> List[T]:
        start_node = self.get_node(start_value)
        end_node = self.get_node(end_value)
        distances: Dict[T, float] = {}
        previous_nodes: Dict[T, T] = {}
        unvisited: Set[GraphNode] = set(self.nodes)

        if start_node:
            distances[start_node.value] = 0.0

        while unvisited:
            current_node = min(unvisited, key=lambda x: distances.get(x.value, float('inf')))
            unvisited.remove(current_node)

            for edge in current_node.edges:
                distance = distances.get(current_node.value, float('inf')) + (edge.weight or 0.0)
                if distance < distances.get(edge.destination.value, float('inf')):
                    distances[edge.destination.value] = distance
                    previous_nodes[edge.destination.value] = current_node.value

        shortest_path = []
        current = end_value
        while current != start_value:
            previous = previous_nodes.get(current)
            if previous is None:
                return []  # No path exists
            shortest_path.insert(0, previous)
            current = previous

        shortest_path.insert(0, start_value)
        return shortest_path


# Usage
if __name__ == "__main__":
    graph = Graph()

    graph.add_node(1)
    graph.add_node(2)
    graph.add_node(3)

    graph.add_edge(1, 2)
    graph.add_edge(2, 3)
    graph.add_edge(3, 1)

    print("Is cyclic:", graph.is_cyclic())
    print("DFS result:", graph.depth_first_search(1))
    print("BFS result:", graph.breadth_first_search(1))
    print("Shortest path from 1 to 3:", graph.shortest_path(1, 3))
